stardis/plasma/molecules.py

```diff
@@ -0,0 +1,422 @@
+import numpy as np
+import pandas as pd
+import logging
+
+from astropy import constants as const, units as u
+from tardis.util.base import element_symbol2atomic_number
+
+from tardis.plasma.properties.base import ProcessingPlasmaProperty
+
+ALPHA_COEFFICIENT = (np.pi * const.e.gauss**2) / (const.m_e.cgs * const.c.cgs)
+
+logger = logging.getLogger(__name__)
+
+
+class MoleculeIonNumberDensity(ProcessingPlasmaProperty):
+    """
+
+    Calculates the number density of each molecule at each depth point using Barklem and Collet 2016 equilibrium constants.
+    Multiplies the number density of each constituent ion and divides by the number density constant. Negative ions are ignored.
+    Attributes
+    ----------
+    molecule_number_density : DataFrame
+            A pandas DataFrame with dtype float. This represents the number density of each molecule at each depth point.
+    molecule_ion_map : DataFrame
+            A pandas DataFrame with the constituent ions of each molecule. Useful for calculating molecular masses later for doppler broadening.
+    """
+
+    # Need to think about negative ions - ignoring for now
+    # applicable for equilibrium constants given by Barklem and Collet 2016, which are given in SI units
+
+    outputs = ("molecule_number_density", "molecule_ion_map")
+
+    def calculate(self, ion_number_density, t_electrons, atomic_data):
+        # Preprocessing - split ions into symbol, charge, and number
+        try:
+            molecules_df = atomic_data.molecule_data.dissociation_energies.copy()
+        except:
+            raise ValueError(
+                "No molecular dissociation energies found in atomic data. Use Carsus to generate atomic data with the Barklem and Collet 2016 data."
+            )
+        molecules_df[
```

Comment:
Perhaps convert to a function that takes `Ion1` or `Ion2` as arguments?

Reply:
Not quite sure what you meant by this but I added a helper function that does the preprocessing that reduces the code duplication a little bit. Let me know if you intended something else.

Reply:
That's perfect